### PR TITLE
Fix issue of initializing hex literals to decimal type

### LINKED ISSUE #14775: Hex floating point literals are always float
**Description:**
Fix $subject in [1] and other places

[1] - tests/ballerina-spec-conformance-tests/S05_simple_values/tests/floating-point-types/decimal.bal

Please refer James response in  https://github.com/ballerina-platform/ballerina-spec/issues/50

### PR DESCRIPTION
**Pull request by chiranSachintha**
Fixes #14775 

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [x] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java [hunk 1]
```diff
@@ -344,7 +344,8 @@ private BType setLiteralValueAndGetType(BLangLiteral literalExpr, BType expType)
             if (expType.tag == TypeTags.FLOAT) {
                 literalType = symTable.floatType;
                 literalExpr.value = ((Long) literalValue).doubleValue();
-            } else if (expType.tag == TypeTags.DECIMAL) {
+            } else if ((expType.tag == TypeTags.DECIMAL) && !NumericLiteralSupport
+                    .isHexIndicator(literalExpr.originalValue)) {
```
Comment (MaryamZi) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java:348]:
```suggestion
            } else if (expType.tag == TypeTags.DECIMAL && 
                    !NumericLiteralSupport.isHexIndicator(literalExpr.originalValue)) {
```

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/util/NumericLiteralSupport.java [hunk 2]
```diff
@@ -68,6 +68,24 @@ public static boolean isHexLiteral(String numericLiteral) {
         return false;
     }
 
+    /**
+     * Check input for decimal indicator.
+     *
+     * @param literalValue literal to check
+     * @return true if Hex prefix is present, false otherwise
+     */
+    public static boolean isHexIndicator(String literalValue) {
```
Comment (MaryamZi) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/util/NumericLiteralSupport.java:77]: Shall we rename this to 
```suggestion
    public static boolean hasHexIndicator(String literalValue) {
```

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/util/NumericLiteralSupport.java [hunk 3]
```diff
@@ -68,6 +68,24 @@ public static boolean isHexLiteral(String numericLiteral) {
...
+        int length = literalValue.length();
+        // There should be at least 3 characters to form hex literal.
+        if (length < 3) {
+            return false;
+        }
+        char firstChar = literalValue.charAt(1);
+        char secondChar = literalValue.charAt(2);
+        return (firstChar == 'x' || secondChar == 'x' || firstChar == 'X' || secondChar == 'X');
```
Comment (MaryamZi) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/util/NumericLiteralSupport.java:85]:
```suggestion
        return firstChar == 'x' || secondChar == 'x' || firstChar == 'X' || secondChar == 'X';
```

Shall we add a comment to say why we are checking both 2nd and 3rd chars?

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java (hunk 1 above)
Comment (chiranSachintha) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java]: Fixed

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/util/NumericLiteralSupport.java (hunk 2 above)
Comment (chiranSachintha) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/util/NumericLiteralSupport.java]: Fixed

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/util/NumericLiteralSupport.java (hunk 3 above)
Comment (chiranSachintha) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/util/NumericLiteralSupport.java]: Fixed

File: tests/jballerina-bstring-unit-test/src/test/resources/test-src/types/decimal/decimal_value_negative_literal.bal [hunk 4]
```diff
@@ -19,6 +19,8 @@ function testInvlaidDecimalValue() {
     // This is to test a syntax issue.
     decimal a = 12.3f;
     decimal b = 0x4fp1;
+    decimal c = 0x12bc;
+    decimal d = -0x12b;
```
Comment (MaryamZi) [inline tests/jballerina-bstring-unit-test/src/test/resources/test-src/types/decimal/decimal_value_negative_literal.bal:23]: Shall we add tests for `0X` and `-0X` too?

Comment (chiranSachintha) [reply tests/jballerina-bstring-unit-test/src/test/resources/test-src/types/decimal/decimal_value_negative_literal.bal:23]: Done

**Pull request merged**